### ISSUE TITLE
How setup enviroment for JUnit tests

### ISSUE THREAD

**Issue by jvaca92** [yidongnan/grpc-spring-boot-starter#225]
Hello,

I would like to have question about how to setup correctly the enviroment for JUnit tests. Actually I have followed some of your examples in folder tests but without success and stuff which are located in folder examples there I cannot find some JUnit tests.
My main goals is to be able used your annotations such as @Grpcservice a @GrpcClient in JUnit test class. For example @GrpcClient always return null if I defined that in JUnit class in case that all needed properties are defined as for client and also for server.

Thank in advance.

**Comment by ST-DDT** [comment]
There are two things you need to keep in mind for tests.

* You have to start a spring-boot context
* You have to enable autoconfiguration or select the required configuration classes yourself.

Here is a simple JUnit5-Test

````java

package net.devh.boot.grpc.test;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;
import org.springframework.boot.SpringBootConfiguration;
import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.test.annotation.DirtiesContext;

import com.google.protobuf.Empty;

import net.devh.boot.grpc.client.inject.GrpcClient;
import net.devh.boot.grpc.test.proto.TestServiceGrpc.TestServiceBlockingStub;
import net.devh.boot.grpc.test.server.TestServiceImpl;

@SpringBootTest(
        properties = {
                "grpc.server.inProcessName=test",
                "grpc.client.test.address=in-process:test"
        })
@DirtiesContext
public class SimpleTest {

    @GrpcClient("test")
    private TestServiceBlockingStub testServiceBlockingStub;

    @Test
    public void simpleTest() {
        assertEquals("1.2.3", this.testServiceBlockingStub.normal(Empty.getDefaultInstance()).getVersion());
    }

    @SpringBootConfiguration // Spring wants one for their tests
    @EnableAutoConfiguration // (potentially slow) Or manually select the relevant classes yourselfs
    // @ImportAutoConfiguration({GrpcServerAutoConfiguration.class, GrpcServerFactoryAutoConfiguration.class, GrpcClientAutoConfiguration.class}) // (minimal set)
    @ComponentScan(basePackageClasses = TestServiceImpl.class) // Could be replaced with a @Bean definition
    public static class Config {
    }

}
````

Does this help you?

**Comment by jvaca92** [comment]
Thanks for quick reply. I tried to modified junit class as you suggest but I am staying at same blocking point.  Here is my code which should be configured same as yours but when do the debugging the UserServiceBlockingStub is every time null. I am not sure if spring really try to find grpc client by annotation @GrpcClient in the junit class.

`//@DataJpaTest
//@AutoConfigureEmbeddedDatabase
@SpringBootTest(
        properties = {
                "grpc.server.inProcessName=test",
                "grpc.client.test.address=in-process:test"
        })
@DirtiesContext
public class GrpcServiceTestCases {

    @SpringBootConfiguration // Spring wants one for their tests
    @EnableAutoConfiguration // (potentially slow) Or manually select the relevant classes yourselfs
    //@ImportAutoConfiguration({GrpcServerAutoConfiguration.class, GrpcServerFactoryAutoConfiguration.class, GrpcClientAutoConfiguration.class}) // (minimal set)
    public static class Config {

        @Bean
        public GrpcUserServiceImpl grpcUserService() {
            return new GrpcUserServiceImpl();
        }
    }

    @GrpcClient("test")
    private UserServiceGrpc.UserServiceBlockingStub userServiceBlockingStub;

//    @Before
//    public void init() {
//        userRepository.save(createUserTest());
//    }

    private UserEntity createUserTest() {
        UserEntity userEntity = new UserEntity();
        userEntity.setUsername("test");
        userEntity.setPassword("test");
        return userEntity;
    }

    /************ Test case *************************/
     /*  Client call for specific user by username
     /*  and remote function return user if exist or
     /*  throw exception if does not exist
     /* **********************************************/
     @Test
     public void callTest_GrpcUserServiceTest() {
         GrpcApplicationUser user = this.userServiceBlockingStub.findUserByUsername(
                 GrpcUsername.newBuilder()
                         .setUsername("test")
                         .build());
         Assert.assertNotNull(user);
     }
}`

**Comment by ST-DDT** [comment]
Can you include the logs for the run with a similar configuration to: https://github.com/yidongnan/grpc-spring-boot-starter/blob/master/tests/src/test/resources/logback-test.xml#L11-L13 ?

Also try adding a break point into [`GrpcClientBeanPostProcessor`](https://github.com/yidongnan/grpc-spring-boot-starter/blob/master/grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/inject/GrpcClientBeanPostProcessor.java)'s constructor and check whether it gets instantiated at all.

**Comment by jvaca92** [comment]
So after debugging and see if the GrpcClientPostProcessor will be invoked I found out that also annotation @RunWith(SpringRunner.class) must be involved then the post processor is invoked. Now I am struggeling with this kind exception which end up that the channel cannot be created.

> 
> java.lang.IllegalStateException: Failed to create channel: test
> 	at net.devh.boot.grpc.client.inject.GrpcClientBeanPostProcessor.processInjectionPoint(GrpcClientBeanPostProcessor.java:118) ~[grpc-client-spring-boot-autoconfigure-2.3.0.RELEASE.jar:2.3.0.RELEASE]
> 	at net.devh.boot.grpc.client.inject.GrpcClientBeanPostProcessor.postProcessBeforeInitialization(GrpcClientBeanPostProcessor.java:77) ~[grpc-client-spring-boot-autoconfigure-2.3.0.RELEASE.jar:2.3.0.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:419) ~[spring-beans-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1737) ~[spring-beans-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:410) ~[spring-beans-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.injectDependencies(DependencyInjectionTestExecutionListener.java:120) ~[spring-test-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.prepareTestInstance(DependencyInjectionTestExecutionListener.java:83) ~[spring-test-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.springframework.boot.test.autoconfigure.SpringBootDependencyInjectionTestExecutionListener.prepareTestInstance(SpringBootDependencyInjectionTestExecutionListener.java:44) ~[spring-boot-test-autoconfigure-2.1.0.RELEASE.jar:2.1.0.RELEASE]
> 	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:246) ~[spring-test-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:227) [spring-test-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:289) [spring-test-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) [junit-4.12.jar:4.12]
> 	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:291) [spring-test-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:246) [spring-test-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97) [spring-test-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
> 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) [junit-4.12.jar:4.12]
> 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) [junit-4.12.jar:4.12]
> 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) [junit-4.12.jar:4.12]
> 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) [junit-4.12.jar:4.12]
> 	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61) [spring-test-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70) [spring-test-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363) [junit-4.12.jar:4.12]
> 	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190) [spring-test-5.1.2.RELEASE.jar:5.1.2.RELEASE]
> 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137) [junit-4.12.jar:4.12]
> 	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68) [junit-rt.jar:na]
> 	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47) [junit-rt.jar:na]
> 	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242) [junit-rt.jar:na]
> 	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70) [junit-rt.jar:na]
> Caused by: java.lang.IllegalArgumentException: java.net.URISyntaxException: Expected scheme name at index 0: :///test
> 	at io.grpc.internal.ManagedChannelImpl.getNameResolver(ManagedChannelImpl.java:648) ~[grpc-core-1.18.0.jar:1.18.0]
> 	at io.grpc.internal.ManagedChannelImpl.<init>(ManagedChannelImpl.java:547) ~[grpc-core-1.18.0.jar:1.18.0]
> 	at io.grpc.internal.AbstractManagedChannelImplBuilder.build(AbstractManagedChannelImplBuilder.java:440) ~[grpc-core-1.18.0.jar:1.18.0]
> 	at net.devh.boot.grpc.client.channelfactory.AbstractChannelFactory.newManagedChannel(AbstractChannelFactory.java:132) ~[grpc-client-spring-boot-autoconfigure-2.3.0.RELEASE.jar:2.3.0.RELEASE]
> 	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660) ~[na:1.8.0_152]
> 	at net.devh.boot.grpc.client.channelfactory.AbstractChannelFactory.createChannel(AbstractChannelFactory.java:98) ~[grpc-client-spring-boot-autoconfigure-2.3.0.RELEASE.jar:2.3.0.RELEASE]
> 	at net.devh.boot.grpc.client.inject.GrpcClientBeanPostProcessor.processInjectionPoint(GrpcClientBeanPostProcessor.java:113) ~[grpc-client-spring-boot-autoconfigure-2.3.0.RELEASE.jar:2.3.0.RELEASE]
> 	... 28 common frames omitted
> Caused by: java.net.URISyntaxException: Expected scheme name at index 0: :///test
> 	at java.net.URI$Parser.fail(URI.java:2848) ~[na:1.8.0_152]
> 	at java.net.URI$Parser.failExpecting(URI.java:2854) ~[na:1.8.0_152]
> 	at java.net.URI$Parser.parse(URI.java:3046) ~[na:1.8.0_152]
> 	at java.net.URI.<init>(URI.java:673) ~[na:1.8.0_152]
> 	at java.net.URI.<init>(URI.java:774) ~[na:1.8.0_152]
> 	at io.grpc.internal.ManagedChannelImpl.getNameResolver(ManagedChannelImpl.java:645) ~[grpc-core-1.18.0.jar:1.18.0]
> 	... 34 common frames omitted

**Comment by ST-DDT** [comment]
Oh yeah, you are right.
The in-process server and client configuration option was added only very recently and is not included in 2.3.0. Sorry i forgot about that.
Then remove and replace both properties with `grpc.client.test.negotitationType=PLAINTEXT`

OR

Define `InProcessGrpcClientFactory` and `InProcessGrpcServerFactory` beans (server name must match client name) in your config and remove the properties .

**Comment by jvaca92** [comment]
Great, I can confirmed that the second solution work. I have pasted here example code how can be managed junit test class if someone would be interested about that. 
Anyway a big thanks for quick help. 
```

package org.maxlot.msa.user.server.test;

import io.grpc.inprocess.InProcessChannelBuilder;
import io.zonky.test.db.AutoConfigureEmbeddedDatabase;
import net.devh.boot.grpc.client.autoconfigure.GrpcClientAutoConfiguration;
import net.devh.boot.grpc.client.channelfactory.GrpcChannelFactory;
import net.devh.boot.grpc.client.channelfactory.InProcessChannelFactory;
import net.devh.boot.grpc.client.config.GrpcChannelsProperties;
import net.devh.boot.grpc.client.inject.GrpcClient;
import net.devh.boot.grpc.client.interceptor.GlobalClientInterceptorRegistry;
import net.devh.boot.grpc.server.autoconfigure.GrpcServerAutoConfiguration;
import net.devh.boot.grpc.server.config.GrpcServerProperties;
import net.devh.boot.grpc.server.serverfactory.GrpcServerFactory;
import net.devh.boot.grpc.server.serverfactory.InProcessGrpcServerFactory;
import net.devh.boot.grpc.server.service.GrpcServiceDefinition;
import net.devh.boot.grpc.server.service.GrpcServiceDiscoverer;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.maxlot.msa.user.api.grpc.GrpcApplicationUser;
import org.maxlot.msa.user.api.grpc.GrpcUsername;
import org.maxlot.msa.user.api.grpc.UserServiceGrpc;
import org.maxlot.msa.user.server.domains.RoleEntity;
import org.maxlot.msa.user.server.domains.UserEntity;
import org.maxlot.msa.user.server.grpc.GrpcUserServiceImpl;
import org.maxlot.msa.user.server.repositories.UserRepository;
import org.maxlot.msa.user.server.test.config.InProcessConfiguration;
import org.maxlot.msa.user.server.test.config.ServiceConfiguration;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.SpringBootConfiguration;
import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.boot.autoconfigure.ImportAutoConfiguration;
import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
import org.springframework.test.context.junit4.SpringRunner;

import java.nio.channels.Channel;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Optional;

/**
 * Implementation of test cases which test functionality of all
 * gRPC services
 * @author Jan Vaca <jan.vaca92@gmail.com> on 4/23/2019
 */
@RunWith(SpringRunner.class)
@SpringBootTest
@DirtiesContext
public class GrpcServiceTestCases {

    @MockBean
    private UserRepository userRepository;

    @SpringBootConfiguration // Spring wants one for their tests
    @EnableAutoConfiguration(exclude = {DataSourceAutoConfiguration.class})
    // (potentially slow) Or manually select the relevant classes yourselfs
    //@ImportAutoConfiguration({GrpcServerAutoConfiguration.class, GrpcServerFactoryAutoConfiguration.class, GrpcClientAutoConfiguration.class}) // (minimal set)
    public static class Config {

        @Bean
        public GrpcUserServiceImpl grpcUserService() {
            return new GrpcUserServiceImpl();
        }

        @Bean
        GrpcChannelFactory grpcChannelFactory(final GrpcChannelsProperties properties,
                                              final GlobalClientInterceptorRegistry globalClientInterceptorRegistry) {
            return new InProcessChannelFactory(properties, globalClientInterceptorRegistry) {

                @Override
                protected InProcessChannelBuilder newChannelBuilder(final String name) {
                    return super.newChannelBuilder("test") // Use fixed inMemory channel name: test
                            .usePlaintext();
                }

            };
        }

        @Bean
        GrpcServerFactory grpcServerFactory(final GrpcServerProperties properties,
                                            final GrpcServiceDiscoverer discoverer) {
            final InProcessGrpcServerFactory factory = new InProcessGrpcServerFactory("test", properties);
            for (final GrpcServiceDefinition service : discoverer.findGrpcServices()) {
                factory.addService(service);
            }
            return factory;
        }
    }

    @GrpcClient("test")
    private UserServiceGrpc.UserServiceBlockingStub userServiceBlockingStub;

    /************ Test case *************************/
     /*  Client call for specific user by username
     /*  and remote function return user if exist or
     /*  throw exception if does not exist
     /* **********************************************/
    @Test
    public void callTest_GrpcUserServiceTest() {
        Mockito.when(userRepository.findById("test")).thenReturn(Optional.of(createTestUser()));
        GrpcApplicationUser user = this.userServiceBlockingStub.findUserByUsername(
                GrpcUsername.newBuilder()
                        .setUsername("test")
                        .build());
        Assert.assertNotNull(user);
    }


    /******************************************
     *         HELPER TEST METHODS
     *****************************************/
    private UserEntity createTestUser() {
        UserEntity userEntity = new UserEntity();
        userEntity.setPassword("test");
        userEntity.setUsername("test");
        userEntity.setActive(true);
        userEntity.setRoles(new HashSet<>());
        return userEntity;
    }
}
```
